### ISSUE TITLE
SPProductUpdate and SPLanguagePack conflict

### ISSUE THREAD

**Issue by NikCharlebois** [PowerShell/SharePointDsc#679]
**Details of the scenario you try and problem that is occurring:**
This was originally reported on the "SharePointDSC Best Practices and Gotchas Post" http://nikcharlebois.com/sharepoint-dsc-gotchas-and-best-practices/

In theory, it should be possible for someone to install, in order, a PU, a Language specific patch, and then run the config wizard. That scenario works just fine if done manually. However, with the SPDSC module, if you specify two SPProductUpdate blocks in a row, and then call the SPConfigWizard, it will skip over the second one because of the way we use B2B and will throw the following message:

Write-Verbose -Message ("An upgrade is pending. " + `
                                "To prevent a possible loop, the install will be skipped")

If this can be done manually, then we should look at having the DSC module allow it as well, given that Language specific patches are now separate from the main CU/PU.

**The DSC configuration that is using the resource:**
N/A

**Version of the Operating System and PowerShell the DSC Target Node is running:**
Win 2016/WMF 5.1

**Version of the DSC module you're using:**
1.9.0.0

**Comment by ykuijs** [comment]
The ProductUpdate resource is using the server version information in the Config database. This information is updated when the Config Wizard is executed. This means that after installing an update, at a next run the resource is unable to check if it already is installed but it is able to detect that the Config Wizard is required.

To prevent a loop (or error messages), the resource does not try to install the update when the Config Wizard is required.

This can be resolved if we find a way to detect the version information locally. So far I haven't got a solution for this problem.

**Comment by NikCharlebois** [comment]
Unless I am mistaken, the [Hive]/ISAPI/Microsoft.SharePoint.dll file will always get updated on a server the moment a CU/PU is applied. We could then check the minor version of that DLL in order to detect if the patch has been applied locally can't we?

**Comment by andikrueger** [comment]
When we try to check the version number of a file, we have to be certain, that this file was part of the patch. If I understand this [post by Stefan Gossner](https://blogs.technet.microsoft.com/stefan_gossner/2016/08/23/sharepoint-does-not-have-a-build-version-full-stop/) correctly, the file versions do not reflect the installed patches.

In cases we install patches manually, there is no check whether an update was installed prior. The updates its self checks, if the update is already in place. See #579. The Updater will exit with an [exit code 17025](https://github.com/PowerShell/SharePointDsc/issues/579#issuecomment-334072305). 
Maybe we can work with this result in the test method or try to get information about the method that is used in the installer.

**Comment by andikrueger** [comment]
I was wondering whats the difference between a manual install of an update and the way we do it with DSC. With SharePointDSC there are assumptions made (Please correct me if I'm wrong):
1. Each ProductUpdate has to be followed by Config Wizard
2. Each Set-TargetResource function of the ProductUpdate Resource checks if there is a pending upgrade. Usually, a user would not do so.
3. Config Wizard can only be used once (properly) in a configuration

Just some ideas on how to resolve this issue:
1. Implement something similar to the Package Resource, that checks via the Package GUID, if the update is installed or not
2. Do not check if there is a pending upgrade; Install the update anyway
3. Try to reinstall the update and use the exit code 17025 to return exit successfully
4. Load all information from the update package and compare it with all dlls in place

**Comment by NikCharlebois** [comment]
@andikrueger, Yes it should be as straight forward as this in theory. The challenge is that simply by looking at the PU/CU package we need to obtain information about what version it is trying to install. Even if we add logic to look inside the package, the .xml file included normally provides this information in the <Sequence> node, but I am not sure if this truly is our best option. 

Also, the bigger challenge here is that unless the file continues to exist, the Test-TargetResource will always return false (also true for language packs). If the user deletes the PU/CU package after installation, then the method can't figure out what version we're trying to install and will throw an error. There are no ways around it.

Unless we add a parameter to the resource to specify what version the SPProductUpdate is for, there's no real way for us to obtain this information from the installer file and then to check to see if it is locally installed.

**Comment by NikCharlebois** [comment]
Our only real option at this point would be to do what you suggested, which is to attempt to install the patch, not matter what, then catch exit code 17025 and threat it as a "Package already installed on server" message. I am just afraid of what that will cause to the performance of the resource. Normally a DSC config should not have more than one SPProductUpdate resource block in it, but some users will not follow best practices and attempt this.

**Comment by ykuijs** [comment]
One of the issues we have here is:
I install SP and the most recent CU. If I then install a Language Pack, I have to reapply the CU in order to have that LP updates to the latest patch level. The current resources are able to do that.

@andikrueger To answer your questions:
1. That is true, but you only need one SPConfigWizard resource in a configuration. That will pick up both Language Pack and CU installs. It does have to be specified early in the configuration, so you are sure it will be executed.
2. Correct, that is because of the way we check the installed updates: Using the Config DB. When an update is installed, but the Config Wizard has not been executed yet, the build numbers in the Config DB are not updated yet. So if SPProductUpdate detects that the Config Wizard is needed, it will refuse to execute to prevent installing the update over and over again.
3. The Config Wizard is only needed in a configuration just once, there is no need to specify it multiple times.

If we can find a way to detect if the update is needed on the environment by not using the Config DB, I am all for it. Unfortunately so far, I haven't found a way yet.

**Comment by andikrueger** [comment]
@NikCharlebois I agree with you. Trying to install the update and wait for the error code is not the best option we have. This will even break the ability to run Test-DscConfiguration as the Test-TargetResource of product update will return false. Also the user has to have the update in place.

@ykuijs Thank you very much for the answers. 

What would happen, if we change the ProductUpdate Resource to something like this:

Parameter | Attribute | DataType | Description | Allowed Values
------------ | ------------- | ------------ | ------------- | -------------
|**Patch** | Key | String | The Name of the patch like: "KB4011244" | All KB patches alailable for SharePoint |
| **SetupFile** | Write | String | The name of the update setup file ||
| **ShutdownServices** | Write | Boolean | Shutdown SharePoint services to speed up installation ||
| **BinaryInstallDays** | Write | String[] | Specify on which dates the installation is allowed |mon, tue, wed, thu, fri, sat, sun|
| **BinaryInstallTime** | Write | String | Specify in which time frame the installation is allowed ||
| **Ensure** | Write | string | Present to install SharePoint. Absent is currently not supported |Present, Absent|
| **InstallAccount** | Write | String | POWERSHELL 4 ONLY: The account to run this resource as, use PsDscRunAsCredential if using PowerShell 5 ||


In case we use a script similar to this [one](https://gallery.technet.microsoft.com/scriptcenter/Get-Installed-Programs-fae091ed), we could  optain all installed patches: 
![image](https://user-images.githubusercontent.com/15608729/33725691-2e67efe2-db73-11e7-9a49-e8ab20840b70.png)

**Comment by ykuijs** [comment]
Using this method has one major issue: When you deployed a CU and then decide to add a Language Pack, the CU (which is multilingual) will never be applied to that language pack.

I have no clue how the CU itself is doing its checks to see if it needs to be installed or not. It might be an idea to reflector the setup file to "borrow" that logic ;-)

**Comment by andikrueger** [comment]
That's something I was already up to. But I was not able to get into the inside of the msiexec file (hope this is the right one). It's not .Net...

The part about the language pack is interesting though. I think I ran into an issue,the moment I tried to reapply a CU after a Language pack. I documented the outcome in #579. Maybe there is more magic in the CU installer or even less magic than we currently can think of. :-)

BTW. What about the SP2016 updates that got split up to start and was files?

**Comment by NikCharlebois** [comment]
Now it has been a while since we revisited this issue, but here are my thoughts. For the Test-TargetResource to work properly, we need to pieces of information:

1 - What is the current patch level on the server (even if not applied to the Config Database with PSConfig yet),  
2 - What is the version that the current patch is trying to apply to our server:

The first piece of information can easily be obtained by calling the following:
(Get-SPDscFarmVersionInfo).Highest

This will return the highest versioned patch installed (even if not committed by psconfig).

The second piece of information can be obtained by extracting the .exe installer, and by looking at the //MSIPatch/SequenceData/Sequence Node value from the sts-x-none.xml file contained within it.

If the second piece of information is greater than the first one, we install the patch, otherwise we assume it is already installed "cumulatively". 

Any reasons issues you folks can think of with this approach?

**Comment by ykuijs** [comment]
Getting the patch level from SP is the way the SPProductUpdate resource currently is doing it, with the exception that I used the Lowest instead of the Highest (Lowest specifies the CU patch level). This is however depending on having the Config Wizard executed.

Getting the patch level from the file is done a little easier in the current resource, by retrieving the product version file property of the setup file.

The challenge is how to retrieve the installed CU on a server, without using the ConfigDB. If we can find a way to get that info from the server instead from the ConfigDB, we are there.

When installing a CU on a system which already has that CU installed, you are prompted that installation isn't necessary. So the patch itself contains somekind of checking mechanism. @andikrueger already tried to reverse engineer the patch, but was unsuccessful so far. I will give it a try as well.

**Comment by andikrueger** [comment]
Thank you for the new thoughts. Unfortunately I haven't come up with any other idea yet. I still hope, we will find a way to mimic the checks made by the installer...
The second part of your solution is promising. We could use that information to come against the current installation. 
I have some thoughts about the first part and hope I'm mistaken:
Isn't Get-SPDscFarmVersionInfo dependent on a Farm in place? If there is this dependence: e.g. we won't be able to install language packs and reapply a CU prior to the farm creation.
In cases a Farm is in place: [SPProductVersions](https://msdn.microsoft.com/en-us/library/office/ee561896.aspx) is obtained from config DB. I haven't tried the command on a SharePoint Server yet. I hope these information get updated by a timer job and not PSConfig (Let's hope for the best!). In case it changes: Your solution looks great.

**Comment by andikrueger** [comment]
@ykuijs Ireally hope, you are lucky with the reverse engineering part of the installer.

**Comment by ykuijs** [comment]
Ok, a quick update: I used ProcessMonitor to check what the install is doing. As it turns out, the installer contains a msp and xml file for each patch. During installation, the package extracts the XML file, reads a GUID from that XML file and check if that GUID is installed.

Unfortunately the GUID is not used exactly as it is stored in the XML file, so have dive into this a little deeper to figure out that one.

Also: I don't think we can reuse this method 100%, since this requires extracting the package each time we validate the resource. But maybe it is an idea to create a script that generates an versioning file based on the package (only once and placed in the same dir as the package) and have the resource use that file. Does that sound as an option?

**Comment by andikrueger** [comment]
Awesome idea with ProcessMonitor!

I really like the idea of the patch version file. This could be a lot of work in the first place to collect all version information fomr SP 2013 and SP 2016 patches, but is well worth it.
With that file in place we are able to use one configuration continually and it would even work in cases, the user [deletes the patch file](https://github.com/PowerShell/SharePointDsc/issues/679#issuecomment-349404370).
What kind of information would you use as key property: The filename of the patch?

Could you give us some more information about the part: "the Guid is not used exactly as stored in the xml file".

**Comment by ykuijs** [comment]
My idea was not to create one patch file which includes the information for all available patches. That is costing too much maintenance on our end. My idea was to create a script (or a method in our module) to have the user create a patch version file as needed. That script extracts the patch, reads all information from the XML files and writes that to the patch file.

During installation, the resource then reads the patch file and checks if each patch is installed on the server.

Looking up the GUID info now......my VM in Azure auto shutdown last night, so lost the information.....have to reproduce the issue ;-)

**Comment by andikrueger** [comment]
@ykuijs That's a good point with the maintenance cost. Why don't we create that file the first time SPProductUpdate or SPLanguagePack are used? This would lower the requirements on the users side and we know for sure the patch-info-file is in place. 

Auto shutdown is sometimes a real pain :)

**Comment by ykuijs** [comment]
Extracting the package can take a very long time (30 minutes or more). Implementing this as part of the resource will introduce a long delay. But of course we can do both: Offer the option to generate the file manually beforehand or generate it when it does not exist.

The GUID in the XML file is: 90150000-115D-042C-1000-0000000FF1CE
The GUID the installer is looking for in the registry is: 00005109D511C2400100000000F01FEC

If you compare these GUIDs, you can see some patterns:
90150000-115D-042C-1000-0000000FF1CE
00005109-D511-C240-0100-000000F01FEC

The first three parts are reversed, in the fourth the first two are reversed (maybe more, but 00 reversed is still 00 ;-)) and in the last part they are also reversed in pairs of two characters.

Right now we have to figure out how to convert the GUID in the file to the GUID used in the registry

**Comment by ykuijs** [comment]
We can use this function: https://gist.github.com/MyITGuy/d3e039c5ec7865edefc157fcd625a20a

**Comment by andikrueger** [comment]
...and here is the description of what is done: [https://blogs.technet.microsoft.com/odsupport/2009/12/17/how-to-convert-an-office-guid-or-office-patch-guid-to-a-compressed-guid/](https://blogs.technet.microsoft.com/odsupport/2009/12/17/how-to-convert-an-office-guid-or-office-patch-guid-to-a-compressed-guid/)

**Comment by NikCharlebois** [comment]
Ok, I've got the code written to obtain the GUID and will commit it into the Util module. Great catch Colombo @ykuijs

**Comment by NikCharlebois** [comment]
We may still have some issues trying to determine if a patch is already installed due to their cumulative nature. For example, I spin off a plain vanilla SP2016 environment, and directly installed the November 2017 CU on it. Based on the XML file, the GUID for that patch is:

    **4F2C8C6B-8A0D-46D9-8D34-C9EEE665B03C**

which will get formatted down to:

    **B6C8C2F4D0A89D64D8439CEE6E560BC3**

Now, I can find a registry key matching that Formatted GUID under

    **HKLM\SOFTWARE\Classes\Installer\Patches**

which tells me it has been installed. However, if after installing the November 2017 CU, I download the April 2017, I need to have a way to determine that it doesn't need to be installed because it has been cumulatively applied by a more recent patch.

The April 2017 formatted patch GUID is:

    **E9AC635FD91CBDA4F9A6E02D1B2E1BBA**

But that is nowhere to be found in the Registry. This means that there are no easy way to figure out if a patch needs to be applied or not given that something more recent may be on the server.

**Comment by andikrueger** [comment]
In that case we would have to stick with a database (patch info file), that contains all information about when a cu was released and the guid (or several guid for each component contained in that cu).

**Comment by andikrueger** [comment]
What happens if you try to install the older cu? Will the installer succeed?

**Comment by NikCharlebois** [comment]
Interesting. With an environment that has the November 2017 CU already installed on it, if I extract the April 2017 CU, and run the extracted .msp, then it installs anyway, and creates a registry entry with the Formatted GUID, even if there is a more recent patch applied.

However, if I run the "unextracted" .exe path, then I get a prompt saying that the patch is already installed. Therefore the .exe is what has the logic to check to see if it is already installed. Maybe an option here is to force the users to extract the CUs (just like we are forcing them to do it for the LP), and have the .msp run no matter what if the associated registry key is not found.

**Comment by andikrueger** [comment]
Are there any files changed when the msp gets installed? If that's the case, we would roll back any newer version of the patched file and have to reapply the newer patch. I think this is the reason why we have to reinstall the current CU when we install any language pack and had a CU already installed. Is this right?

**Comment by ykuijs** [comment]
A CU can contain hunderds of MSP's (in SP2013 anyways). So I don't like forcing the user to extract the CU package.

I have analyzed the ProcMon trace a little more and with a patch that is not required, the logic checks several reg keys which all do not exist.
When a patch is required, is checking more regkeys, including a DisplayVersion key. The XML file that is supplied with the MSP also contains a version number and I believe the logic is comparing these versions.

The logic also does some other checks, but I am not sure what theses are for yet.

**Comment by andikrueger** [comment]
Today I spent some time analyzing two SharePoint 2016 patches, that are not needed on my system. I did the following:

- Compared procom traces of both setups.
- Checked the opatchinstall.log files
- the XML file corresponding to a patch
- tried to understand the logic that is used.

What are my findings with the following file: sts2016-kb4011217-fullfile-x64-glb.exe

1. Double Click the .exe
2. Accept the Eula
3. Next
will lead to this screen:
![image](https://user-images.githubusercontent.com/15608729/36477082-40563088-1700-11e8-953b-7589237d0cd7.png)

After reviewing the opatchinstall.log I could trace back, what the installer does:

After my step 1 (minimized log output)
```
OPatchInstall: Will extract the file with ID '$(PATCH.FILEID.EULA)' to the path '$(PACKAGE.PATH.TEMP)\eula.txt'
...
OPatchInstall: Will extract the file with ID 'supplementaleula_sharepoint_en-us.txt' to the path 'C:\Users\admak\AppData\Local\Temp\OWPA29D.tmp\eula.txt'
...
OPatchInstall: Reading EULA from file 'C:\Users\admak\AppData\Local\Temp\OWPA29D.tmp\eula.txt'
```
After my step 2:
```
OPatchInstall: Will extract the file with ID 'sts-x-none.xml' to the path '$(PACKAGE.EXTRACT.FOLDER)\sts-x-none.xml'
...
OPatchInstall: Adding the patch '$(PACKAGE.PATH.TEMP)\sts-x-none.msp' to the queue of patches to install
OPatchInstall: Will log the installation to '$(parentPath:LOGPATH)\sts-x-none.LOG'
OPatchInstall: Will use the command line '$(PACKAGE.PATCHES.DEFAULT.ARGUMENTS) $(PATCH_ARGS)'
...

```
When the .xml file is present the registry will be checked for the following registry keys, which are partially obtained from the xml file (contained in the patch):
```
<MsiPatch xmlns="http://www.microsoft.com/msi/patch_applicability.xsd" SchemaVersion="1.0.0.0" PatchGUID="{358FACB8-17F1-4C89-A9BC-5CD2C479B53C}" MinMsiVersion="5">
	<TargetProduct MinMsiVersion="405">
		<TargetProductCode Validate="true">{90160000-1014-0000-1000-0000000FF1CE}</TargetProductCode>
		<TargetVersion Validate="true" ComparisonType="Equal" ComparisonFilter="MajorMinorUpdate">16.0.4351.1000</TargetVersion>
		<TargetLanguage Validate="false">0</TargetLanguage>
		<UpdatedLanguages>0</UpdatedLanguages>
		<UpgradeCode Validate="true">{00160000-1014-0000-1000-0000000FF1CE}</UpgradeCode>
	</TargetProduct>
	<TargetProductCode>{90160000-1014-0000-1000-0000000FF1CE}</TargetProductCode>
	<SequenceData>
		<PatchFamily>sts</PatchFamily>
		<Sequence>16.0.4600.1002</Sequence>
		<Attributes>1</Attributes>
	</SequenceData>
</MsiPatch>
```
The TargetProductCode `90160000-1014-0000-1000-0000000FF1CE` will lead to this reg key: 
![image](https://user-images.githubusercontent.com/15608729/36479440-04c7e8b4-1709-11e8-9f6b-cf2949329e91.png)
There is a folder Patches:
![image](https://user-images.githubusercontent.com/15608729/36479451-104214a8-1709-11e8-811b-d95e3f562dd7.png)
With the Name of that folder we can navigate to the patches folder with the same name:
![image](https://user-images.githubusercontent.com/15608729/36479493-320e1492-1709-11e8-88c8-e8b098171655.png)
and get the path to the local .msp file of the current patch installed: `C:\Windows\Installer\ea43.msp`

This msp file contains the needed information to compare the current PatchSequence to the patch installed:
Patch we try to install:
```
	<SequenceData>
		<PatchFamily>sts</PatchFamily>
		<Sequence>16.0.4600.1002</Sequence>
		<Attributes>1</Attributes>
	</SequenceData>
```
Path that is installed:
![image](https://user-images.githubusercontent.com/15608729/36479585-81bbdd44-1709-11e8-9c86-6d71764755c0.png)
With that information the installer knows whether to install or not to install. 

This means we have to look into the patch, open the registry, get the installed patch GUID, open the patches folder and compare the PathSequences.

**Comment by andikrueger** [comment]
Following my last post, I will start to investigate if the patch version wich is going to be applied can be obtained from the .exe file information. If so, and if this version number matches each of the contained patches, we would be able to determine if an update is necessary or not, by just comparing the version against the version in the registry.

The thing I did not verify yet: what about language packs that need to be patched. Is it the same procedure or not. 

@NikCharlebois @ykuijs any further ideas what to look for?

**Comment by andikrueger** [comment]
Linking  #708 to this issue.

**Comment by andikrueger** [comment]
I did some more testing to verify my previous post.

The Farms latest patch is the Januar CU 2018 (16.0.4639.1002). Today I installed the Feb 2018 CU. In the following screenshot are some more details on how the patch installer verifies if the patch is installed:
![image](https://user-images.githubusercontent.com/15608729/36631470-9aaf1afc-1978-11e8-8597-2dc4a3e7e09e.png)

1. The product code
2. The installer of the previous patch (see [above](https://github.com/PowerShell/SharePointDsc/issues/679#issuecomment-367308556)
3. The current patch and its GUID + Patch Order
4. The previous patch and the well formatted GUID (As Compressed GUID: 22C0E2B49ABF8A24A82FD7B8059314FD) 

This is the patch data of the Feb 2018 CU:
```
<SequenceData>
		<PatchFamily>sts</PatchFamily>
		<Sequence>16.0.4654.1000</Sequence>
		<Attributes>1</Attributes>
	</SequenceData>
```
The Sequence or version number above equals the one in the file information dialog:
![image](https://user-images.githubusercontent.com/15608729/36631501-62008140-1979-11e8-89d9-53d33461a42f.png)

I see the following ways to determine if we have to apply a patch or not:
1. The filenames for SP 2016 Patches do start with sts2016 (language independent) or wssloc2016 (language dependant). 
a. STS patches are targeting the following product GUID 
`<TargetProductCode Validate="true">{90160000-1014-0000-1000-0000000FF1CE}</TargetProductCode>`
b. WSSLOC: The GUID for the Target Product varies with every language. The installed languages can be obtained from:`
[HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Shared Tools\Web Server Extensions\16.0\InstalledLanguages]
"1033"="16.0.4351.1000"
"1031"="16.0.4351.1000"
`
I did not find a way to get either the GUID or the compressed GUID for the Language Packs. Even with an English installation there is an Englisch Language Pack in place. Currently I only see the option to save all compressed or well formatted GUIDs for the language packs.
2. With the information about the registry keys we can now look for the latest patch and its Sequence.
3. Compare that sequence to the one present in the file information.
a. If the current is lower: Install the patch
b. If they match or the one in the patch  is lower: Do nothing.

**Comment by andikrueger** [comment]
Currently there is still one big challenge left. For SharePoint 2016 there are 49 Language Packs and therefor we would have to know every product guid to find the right registry keys. 
By now I was not able to find a way on how we could determine the right registry keys. I see the following ways to solve this shortcoming:
1. Something in brute force style: @NikCharlebois has something in the SharePointDSC.Reverse Script that could be very useful to finde every installed language pack. the SharePointDSC.Reverse Script uses the [uninstall Information](https://github.com/Microsoft/SharePointDSC.Reverse/blob/43ab5422a9334a5a4678cda7ccfa74a78137420b/SharePointDSC.Reverse.ps1#L663) to gather the installed components. 
2. We could store every Language Pack Guid in our resources. This means we would have to maintain this information too.

I tend to [1] and would start with a draft implementation based on the findings above. Please let me know, if you see any other way.

**Comment by NikCharlebois** [comment]
@andikrueger Great job on this by the way buddy! I tend to prefer option 1 as well, however we need to build an approach that will also support whatever comes out for SP2019 at this stage of the game. We do have a commitment to support SP2019 from Day 1. I do believe the GUID format will remain the same for the 2019 wave, but it's something we will have to verify internally first. So in the meantime, I say it's safe to use the Registry RegEx approach.

**Comment by andikrueger** [comment]
There is some news:
I did more reverse engineering on the update process and am very confident to have a solution for this ready soon.

Currently there is one big constraint, I did not see coming... For SharePoint 2013 there are security updates - how did I forget about these??? 
For SharePoint 2016 there are not.

The SharePoint 2013 Security Updates are named like "wdsrvloc2013", "coreserverloc2013" and so on. They patch a single SharePoint part and do have an own product code. 
Currently I'm stuck at this point. I do not want to store every possible product code, but I can't think of a way, that does not need, to extract the patch file.

**Comment by andikrueger** [comment]
@ykuijs could you assign this topic to me and add an enhancement label? I'm still working on it...

**Comment by ykuijs** [comment]
@andikrueger Since you do not have the correct permissions level in this repo, I cannot assign the issue to you. But it is all yours! Please also checkout #839, since this might have a relation with this issue.

**Comment by andikrueger** [comment]
@ykuijs I updated my branch for this issue with the latest v3.0 changes and will add unit testing fpr the current changes. 
Parts of issue #839 should be fixed by the new  `Get-SPDscLocalVersionInfo` function as we are evaluating the state of a single server and not the farm.

**Comment by andikrueger** [comment]
I'm done with integration testing on SharePoint Server 2016 - Why are SharePoint update THAT time consuming? 

The updated mechanism does a great job. It's possible to leave all prior SPProductUpdates and SPInstallLanguagePack Resources in a configuration :-) 

Please find attached my test configuration and the log of the DSC configuration.
[SPProductUpdateTestConfiguration.txt](https://github.com/PowerShell/SharePointDsc/files/2469051/SPProductUpdateTestConfiguration.txt)
[log.txt](https://github.com/PowerShell/SharePointDsc/files/2469048/log.txt)

**Comment by andikrueger** [comment]
Today Pester got updated with the ability to add a TestRegistry: https://github.com/pester/Pester/pull/1112 This should make unit testing of this resource easier.

I'm looking forward to update my recent PR.

**Comment by andikrueger** [comment]
Yesterday a new version of Pester got released, that allows the usage of TestRegistry:\ https://github.com/pester/Pester/releases/tag/4.5.0

**Comment by ykuijs** [comment]
Will be fixed by PR #1046

**Comment by ykuijs** [comment]
Closed via PR #1046 

A huge thanks to @andikrueger for his work on investigating and implementing this issue!!!
